File: server/app/models/watchlist.py
"""Watchlist sharing models — Supabase-backed persistent storage.

Implements Spotify-style public/private visibility.  Only watchlists
explicitly marked as ``public`` by the user can be shared.
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class WatchlistShareItem(BaseModel):
    """A single item within a shared watchlist."""
    id: str = Field(..., max_length=100)
    saved_title: str = Field(..., min_length=1, max_length=200)
    movie: dict  # Full MovieData as dict (flexible schema)
    added_at: str = Field(..., max_length=50)


class WatchlistShareRequest(BaseModel):
    """Request body for creating a shared watchlist."""
    folder_name: str = Field(..., min_length=1, max_length=100)
    folder_icon: str | None = Field(default=None, max_length=50)
    items: list[WatchlistShareItem] = Field(..., min_length=1, max_length=500)
    visibility: str = Field(default="public", pattern=r"^(public|private)$")  # must be "public" to share


class WatchlistShareResponse(BaseModel):
    """Response after creating a share link."""
    ok: bool = True
    share_token: str
    share_url: str
    visibility: str = "public"
    item_count: int = 0
    created_at: str = ""


class SharedWatchlistView(BaseModel):
    """Public view of a shared watchlist (no auth required to view)."""
    folder_name: str
    folder_icon: str | None = None
    items: list[WatchlistShareItem] = []
    shared_by: str = ""
    created_at: str = ""
    item_count: int = 0
    visibility: str = "public"

File: mobile/lib/models/watchlist.dart
// Watchlist data models (mirrors Supabase schema).

import 'movie_data.dart';

class WatchlistFolder {
  final String id;
  final String name;
  final String? icon;
  final bool isPublic;
  final List<WatchlistItem> items;

  WatchlistFolder({
    required this.id,
    required this.name,
    this.icon,
    this.isPublic = false,
    required this.items,
  });

  factory WatchlistFolder.fromJson(Map<String, dynamic> json) {
    return WatchlistFolder(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      icon: json['icon'],
      isPublic: json['is_public'] ?? false,
      items: (json['items'] as List?)
              ?.map((i) => WatchlistItem.fromJson(i as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'icon': icon,
        'is_public': isPublic,
        'items': items.map((i) => i.toJson()).toList(),
      };
}

class WatchlistItem {
  final String id;
  final String savedTitle;
  final MovieData movie;
  final String addedAt;

  WatchlistItem({
    required this.id,
    required this.savedTitle,
    required this.movie,
    required this.addedAt,
  });

  factory WatchlistItem.fromJson(Map<String, dynamic> json) {
    return WatchlistItem(
      id: json['id'] ?? '',
      savedTitle: json['saved_title'] ?? '',
      movie: MovieData.fromJson(json['movie'] as Map<String, dynamic>),
      addedAt: json['added_at'] ?? '',
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'saved_title': savedTitle,
        'added_at': addedAt,
      };
}
